control-plane/agents/src/bin/core/volume/service.rs

```diff
@@ -426,6 +447,21 @@ impl Service {
             }
             Err(error) => Err(error),
         }?;
+
+        let max_snapshots = if let Some(max_limit) = volume.as_ref().max_snapshots {
+            max_limit
+        } else {
+            self.registry.max_snapshots()
+        };
+        if max_snapshots == 0
+            || (max_snapshots > 0
+                && volume.as_ref().metadata.num_snapshots() as i32 >= max_snapshots)
```

Comment:
Doesn't this cover it?

```suggestion
        if volume.as_ref().metadata.num_snapshots() as i32 >= max_snapshots)
```